src/coprocessor/statistics/histogram.rs

```diff
@@ -0,0 +1,233 @@
+// Copyright 2017 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// FIXME: remove following later
+#![allow(dead_code)]
+
+use protobuf::RepeatedField;
+use tipb::analyze;
+
+/// Bucket is an element of histogram.
+struct Bucket {
+    // the number of items stored in all previous buckets and the current bucket.
+    count: u64,
+    // the greatest item value stored in the bucket.
+    upper_bound: Vec<u8>,
+    // the lowest item value stored in the bucket.
+    lower_bound: Vec<u8>,
+    // the number of repeats of the bucket's upper_bound, it can be used to find popular values.
+    repeats: u64,
+}
+
+impl Bucket {
+    fn new(count: u64, upper_bound: Vec<u8>, lower_bound: Vec<u8>, repeats: u64) -> Bucket {
+        Bucket {
+            count: count,
+            upper_bound: upper_bound,
+            lower_bound: lower_bound,
+            repeats: repeats,
+        }
+    }
+
+    fn append_repeated_item(&mut self) {
+        self.count += 1;
+        self.repeats += 1;
+        return;
+    }
+
+    // insert a item bigger than current upper_bound,
+    // we need data to set as upper_bound
+    fn append_diff_item(&mut self, data: Vec<u8>) {
```

Comment:
I propose to keep one `append` method, and return `Some(data)` if it's failed to be appended due to full.

Reply:
The bucket can not check whether it's full since it only save the number of items which are smaller than or equal to the bucket's `upper_bound`